File: backend/src/routes/evaluations.ts
import { Router, Response } from 'express';
import { supabase } from '../utils/supabase';
import { AuthRequest, requireRole } from '../middleware/auth';

const router = Router();

/**
 * List evaluations with role-based access
 * - ADMIN/PM/RSP/LND: Can see all evaluations
 * - RATER/INTERVIEWER: Can see evaluations they created
 */
router.get('/', requireRole('ADMIN', 'PM', 'RSP', 'LND', 'RATER', 'INTERVIEWER'),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'Not authenticated' });
        return;
      }

      const { applicant_id } = req.query;

      let query = supabase.from('evaluations').select('*');

      if (applicant_id) {
        query = query.eq('applicant_id', applicant_id as string);
      }

      if (req.user.role === 'RATER' || req.user.role === 'INTERVIEWER') {
        // Can only see their own evaluations
        query = query.eq('evaluator_id', req.user.userId);
      }

      const { data, error } = await query.order('created_at', { ascending: false });

      if (error) {
        res.status(500).json({ error: error.message });
        return;
      }

      res.json(data);
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  }
);

/**
 * Create an evaluation (Rater/Interviewer only)
 * Each user can only evaluate applicants assigned to them
 */
router.post('/', requireRole('RATER', 'INTERVIEWER'),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'Not authenticated' });
        return;
      }

      const { applicant_id, score, comments } = req.body;

      if (!applicant_id || score === undefined) {
        res.status(400).json({ error: 'applicant_id and score are required' });
        return;
      }

      if (score < 0 || score > 100) {
        res.status(400).json({ error: 'Score must be between 0 and 100' });
        return;
      }

      // TODO: Add validation to ensure user is assigned to this applicant

      const evaluationData = {
        applicant_id,
        evaluator_id: req.user.userId,
        score,
        comments: comments || null,
      };

      const { data, error } = await supabase
        .from('evaluations')
        .insert(evaluationData)
        .select()
        .single();

      if (error) {
        res.status(500).json({ error: error.message });
        return;
      }

      res.status(201).json(data);
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  }
);

export default router;

File: backend/app/routes/evaluations.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from typing import List, Optional
from app.models.user import UserRole
from app.core.supabase_client import db
from app.utils.dependencies import get_current_user, require_role
from pydantic import BaseModel

router = APIRouter(prefix="/api/evaluations", tags=["evaluations"])


class EvaluationCreate(BaseModel):
    applicant_id: str
    score: float
    comments: Optional[str] = None


class EvaluationResponse(BaseModel):
    id: str
    applicant_id: str
    evaluator_id: str
    score: float
    comments: Optional[str]


@router.get("/", response_model=List[EvaluationResponse])
async def list_evaluations(
    current_user: UserRole = Depends(require_role("ADMIN", "PM", "RSP", "LND", "RATER")),
    applicant_id: Optional[str] = Query(None),
):
    """
    List evaluations with role-based access:
    - ADMIN/PM/RSP/LND: Can see all evaluations
    - RATER/INTERVIEWER: Can see evaluations they created
    """
    try:
        client = db.get_client()
        query = client.table("evaluations").select("*")

        if applicant_id:
            query = query.eq("applicant_id", applicant_id)

        if current_user.role in ["RATER", "INTERVIEWER"]:
            # Can only see their own evaluations
            query = query.eq("evaluator_id", current_user.user_id)

        response = query.execute()
        return response.data

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch evaluations: {str(e)}",
        )


@router.post("/", response_model=EvaluationResponse)
async def create_evaluation(
    evaluation: EvaluationCreate,
    current_user: UserRole = Depends(require_role("RATER", "INTERVIEWER")),
):
    """
    Create an evaluation (Rater/Interviewer only).
    Each user can only evaluate applicants assigned to them.
    """
    try:
        client = db.get_client()

        # TODO: Add validation to ensure user is assigned to this applicant

        evaluation_data = {
            "applicant_id": evaluation.applicant_id,
            "evaluator_id": current_user.user_id,
            "score": evaluation.score,
            "comments": evaluation.comments,
        }

        response = client.table("evaluations").insert(evaluation_data).execute()
        return response.data[0]

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create evaluation: {str(e)}",
        )
